### ISSUE TITLE
@BeforeSuite and @AfterSuite run multiple times when executing tests in parallel

### ISSUE THREAD

**Issue by ryudice** [getgauge/gauge-java#297]
### Expected behavior
When running threads in parallel @BeforeSuite and @AfterSuite should run only once

### Actual behavior
@BeforeSuite and @AfterSuite execute each time for each parallel stream

### Steps to reproduce
Jun add @AfterSuite and run tests in parallel, AfterSuite method will run multiple times

### Gauge version
```
1.0.4
```

**Comment by negiDharmendra** [comment]
@ryudice As of now each thread of the Java Runner is considered as a separate runner by Gauge that's why the  `BeforeSuite` and `AfterSuite` runs for each thread.

**Comment by ryudice** [comment]
@negiDharmendra and is that correct? at the end you are running a Suite, the fact that the suite is run in parallel is a low level abstraction that I dont think you should need to care at the high level. Also the naming of the attribute is misleading if that's the case.

Is there an alternative attribute that gets ran after the executor finishes running all parallel streams?

**Comment by nehashri** [comment]
@ryudice Gauge tries to divide a test suite into multiple parallel suites while running in parallel. This is so that the setup can be done on each of the runner processes before execution. This behaviour is true for multithreaded execution too. I agree that this becomes confusing while using multithreading with Gauge Java.

> Is there an alternative attribute that gets ran after the executor finishes running all parallel streams?
This doesn't exist currently, however if you think this is useful, please create feature request for the same along with your thoughts. This is probably the correct way to fix your problem.

As of now, you may have to find a workaround for this. You can try writing custom code to execute the hook only on the last thread executing the `@AfterSuite` hook.

**Comment by jonathangiber** [comment]
Maybe this one can solve the problem https://github.com/getgauge/gauge/issues/805

**Comment by Vishnu1067** [comment]
+1

Would be nice to have this feature. 

Example: We have one suite which has 30 test cases (runs in parallel) and First step is creating a user. Right now we have this step for all 30 cases and each takes around 5~10 secs. If there could be possible to create @BeforeSuite and store the value in DataStore suite level, can share this value for all other 29 test cases.

**Comment by olejkavn** [comment]
Hi! any updates about this point? 
I have java+spring project 
My use case for this stuff it's spring context initialization
Now I move my context initialization to static block.
But in this case Context start twice.
1st time on gauge validation phase 
2nd during test execution

It's produce situation when I can't scale my test count without gauge timeouts change

**Comment by sriv** [comment]
This is a bug, unfortunately due to the way gauge is designed (see [comment](https://github.com/getgauge/gauge-java/issues/297#issuecomment-527741315)). 

This isn't the way it should be, but it will require some work to sort this out. So far the biggest limitation has been that Gauge supports parallelization by spinning up multiple runner instances. 

We will be picking this up.

**Comment by olejkavn** [comment]
@sriv maybe you have some kind of workaround for this stuff? like some property or somewhere else

**Comment by sriv** [comment]
Nothing that i can think of, i'm afraid. This will require some code changes.

**Comment by sethisandeepsethi** [comment]
Hi,
Can anyone provide some workaround to this problem in Java.
I want to send an email with test summary after execution of the suite. We are dependent on @AfterSuite method for this.
Currently the email is getting sent that many number of time as we have test cases.

**Comment by sethisandeepsethi** [comment]
Hi,
Can anyone provide some workaround to this problem in Java.
I want to send an email with test summary after execution of the suite. We are dependent on @AfterSuite method for this.
Currently the email is getting sent that many number of time as we have test cases.

**Comment by olejkavn** [comment]
@sethisandeepsethi do you use latest versions of gauge? this issue is closed

**Comment by ccrj1105** [comment]
@sethisandeepsethi  I have the same problem with gauge in version 1.1.1

**Comment by sriv** [comment]
@sethisandeepsethi @ccrj1105 - the `BeforeSuite` and `AfterSuite` hooks should run only once, provided you have `enable_multithreading` set to true (i.e. your tests run in multi thread mode).

Note that this may need your tests to be thread safe.

**Comment by bansal63** [comment]
@sriv 
I'm also facing same problem.

I've added a post https://spectrum.chat/gauge/general/can-we-run-specs-in-parallel-when-we-have-data-driven-specifications~c97f23ed-fea9-42fa-b454-7cadadcbaa5b as well.

**Comment by zabil** [comment]
@bansal63 Please share a sample project to replicate this.
